flink-yarn/src/test/java/org/apache/flink/yarn/ResourceInformationReflectorTest.java

```diff
@@ -0,0 +1,173 @@
+		final ResourceWithMethod resourceWithMethod = new ResourceWithMethod();
+		resourceInformationReflector.setResourceInformationUnSafe(resourceWithMethod, RESOURCE_NAME, RESOURCE_VALUE);
+
+		assertNotNull(resourceWithMethod.getResourceWithName(RESOURCE_NAME));
+		assertThat(resourceWithMethod.getResourceWithName(RESOURCE_NAME).getName(), is(RESOURCE_NAME));
+		assertThat(resourceWithMethod.getResourceWithName(RESOURCE_NAME).getValue(), is(RESOURCE_VALUE));
+	}
+
+	@Test
+	public void testGetResourceInformationIfMethodPresent() {
+		final ResourceInformationReflector resourceInformationReflector = new ResourceInformationReflector(ResourceWithMethod.class.getName(), ResourceInfoWithMethod.class.getName());
+		final ResourceWithMethod resourceWithMethod = new ResourceWithMethod();
+		resourceWithMethod.setResourceInformation(RESOURCE_NAME, ResourceInfoWithMethod.newInstance(RESOURCE_NAME, RESOURCE_VALUE));
+
+		final Map<String, Long> externalResources = resourceInformationReflector.getExternalResourcesUnSafe(resourceWithMethod);
+		assertThat(externalResources.size(), is(1));
+		assertTrue(externalResources.containsKey(RESOURCE_NAME));
+		assertThat(externalResources.get(RESOURCE_NAME), is(RESOURCE_VALUE));
+	}
+
+	@Test
+	public void testSetResourceInformationIfMethodAbsent() {
+		final ResourceInformationReflector resourceInformationReflector = new ResourceInformationReflector(ResourceWithoutMethod.class.getName(), ResourceInfoWithMethod.class.getName());
+		final ResourceWithMethod resourceWithMethod = new ResourceWithMethod();
+		resourceInformationReflector.setResourceInformationUnSafe(resourceWithMethod, RESOURCE_NAME, RESOURCE_VALUE);
+
+		assertNull(resourceWithMethod.getResourceWithName(RESOURCE_NAME));
+	}
+
+	@Test
+	public void testGetResourceInformationIfMethodAbsent() {
+		final ResourceInformationReflector resourceInformationReflector = new ResourceInformationReflector(ResourceWithoutMethod.class.getName(), ResourceInfoWithMethod.class.getName());
+		final ResourceWithMethod resourceWithMethod = new ResourceWithMethod();
+		resourceWithMethod.setResourceInformation(RESOURCE_NAME, ResourceInfoWithMethod.newInstance(RESOURCE_NAME, RESOURCE_VALUE));
+
+		final Map<String, Long> externalResources = resourceInformationReflector.getExternalResourcesUnSafe(resourceWithMethod);
+		assertThat(externalResources.entrySet(), is(empty()));
+	}
+
+	@Test
+	public void testSetAndGetExtendedResourcesWithYarnSupport() {
+		assumeTrue(HadoopUtils.isMinHadoopVersion(2, 10));
+
+		final Resource resource = Resource.newInstance(100, 1);
+
+		ResourceInformationReflector.INSTANCE.setResourceInformation(resource, RESOURCE_NAME, RESOURCE_VALUE);
+
+		final Map<String, Long> externalResourcesResult = ResourceInformationReflector.INSTANCE.getExternalResources(resource);
+		assertThat(externalResourcesResult.size(), is(1));
```

Comment:
Shouldn't this have size `2` for `cpu` and `mem`? Otherwise the starting index of the loop in `ResourceInformationReflector. getExternalResourcesUnSafe` does not make sense.

I would suggest to run these tests with Hadoop 2.10 and Hadoop 3 in order to make sure that they are actually running.

Reply:
Do you mean to run Flink's CI test with Hadoop 2.10+ or 3.0+? I think it would be good. Maybe we need to discuss in dev ML?

Reply:
I think the value should be 1 since we ignore the first two elements in `ResourceInformationReflector. getExternalResourcesUnSafe`.

Reply:
You are right. My bad.

Concerning running the tests, could we run the tests with Hadoop 2.10 and 3 locally just for verification purposes?

Reply:
Yes, I'll do it. Actually, I've tested it with Hadoop 2.10(not the UT, but with real GPU driver).

Reply:
Cool :-)

Reply:
After a deeper investigation here. I think this test would be valid.
First, Hadoop only accept the resource type which known to Hadoop, from 3.1.0 and 2.10.0, it is defined in [ResourceInformation.java#L56](https://github.com/apache/hadoop/blob/e9aa1789c2c75fd20600d2a0d11774161319af2a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-api/src/main/java/org/apache/hadoop/yarn/api/records/ResourceInformation.java#L56). It means probably we could only set memory and cpu in 3.0.0, although the mechanism to set the external resource is available.
Meanwhile, it needs to add the specific external resource to Yarn config file.

I think it probably makes sense to just test if there are two default resource types(cpu and memory) in the `Resource` here.

Reply:
Sounds good to me.